CONTRIBUTING.md

```diff
@@ -1,79 +1,109 @@
+**Table of Contents**
 
-## Commit Message Format
+- [Getting Started](#getting-started)
+  - [Prerequisites](#prerequisites)
+  - [Notes](#notes)
+  - [Build](#build)
+  - [Start](#start)
+  - [Run](#run)
+- [Workflow](#workflow)
+  - [Coding Guidelines](#coding-guidelines)
+  - [Scope](#scope)
+- [Thank you](#thank-you)
 
-We follow [Angular's Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit)
+## Getting Started
 
-Each commit message consists of a **header**, a **body** and a **footer**. The header has a special format that includes a **type**, a **scope** and a **subject**:
+If you wish to better understand how Viewer works or want to debug an issue: get the source, build it, and run it locally.
 
+### Prerequisites
+
+In order to download necessary tools, clone the repository, and install dependencies, you'll need network access.
+
+You'll need the following:
+
+- [Git](https://git-scm.com/)
+- [Node.JS](https://nodejs.org) (If you need to manage multiple Node.JS versions, [use a node version manager](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) to install)
+
+Get the source code of `viewer` repo:
+
+```shell
+gh repo fork shabados/viewer --clone=true
 ```
-<type>(<scope>): <subject>
-<BLANK LINE>
-<body>
-<BLANK LINE>
-<footer>
-```
 
-The **header** is mandatory and the **scope** of the header is optional.
+**PROTIP**: Use the [`gh` cli tool from GitHub](https://cli.github.com/) to fork the repo to your GitHub account (if not already), clone it to your local machine, and set the appropriate remotes for origin and upstream with the above command.
+
+### Notes
+
+You will need to create a Google Cloud Project with the Translate API enabled to try out punjabi to english translating.
+
+Translations will be enabled if the `GOOGLE_APPLICATION_CREDENTIALS` environment variable is set to a path that contains the credentials for a Google Cloud Project's service account, which you must create. [Learn more about setting up the Google Cloud Project](https://www.npmjs.com/package/@google-cloud/translate#quickstart).
 
-Any line of the commit message cannot be longer 100 characters! This allows the message to be easier to read on GitHub as well as in various git tools.
+### Build
 
-The footer should contain a [closing reference to an issue](https://help.github.com/articles/closing-issues-via-commit-messages/) if any.
+Run `npm i` in the root directory.
 
-Samples:
+### Start
 
-```docs(changelog): update changelog to beta.5```
+```shell
+npm start
+```
+
+### Run
+
+Usage:
 
+```shell
+npm run <command>
 ```
-fix(release): need to depend on latest rxjs and zone.js
 
-The version in our package.json gets copied to the one we publish, and users need the latest of these.
+The commands are:
+
+```shell
+start:backend          # Start backend
+start:frontend         # Start frontend
+build                  # Build frontend
+install-deps           # Run npm install for entire project
```

Comment:
This script will install dependencies for `frontend/` and `backend/` but for this to work we need the top-level dependencies to be installed. I think we should just remove this one since this is just a helped that runs as post-install script.

Reply:
I think you should create a new issue/PR for that then -- and update the contributing docs while you do that

Does that sound good @Harjot1Singh ?

Reply:
Rest looks good to me. I just meant to remove `install-deps` from this section because of the reason mentioned. I selected a line but it gave me 4 lines and that is confusing.

Just making this change would suffice imo
```suggestion

```

Reply:
Why does the command exist?

Reply:
Following is structure of this project.
```
.
├── CONTRIBUTING.md
├── LICENSE.md
├── README.md
├── docker-compose.yml
├── backend 
│   ├── Dockerfile
│   ├── api.js
│   ├── db.js
│   ├── package-lock.json
│   ├── package.json
│   ├── server.js
│   └── translate.js
├── docs
├── frontend
│   ├──  Dockerfile
│   ├── package-lock.json
│   ├── package.json
│   ├── public
│   └── src
├── package-lock.json
├── package.json
└── repologo.png
```
When we do npm install at `.` we are installing all the dependencies and dev dependencies in `./package.json`. To install the dependencies and dev dependencies for `./frontend//package.json` or `./backend/package.json` you have to `cd` into that directory. To make it easy we have a `install-deps:backend` and `install-deps:frontend` in `./package.json` file which changes directory and installs the deps.

NPM allows you to run [pre and post scripts](https://docs.npmjs.com/cli/v6/using-npm/scripts#pre--post-scripts). We are using a post install script, once you do `npm i` in `.` we are essential doing `npm run install-deps:backend && install-deps:frontend` in our`postinstall` scripts. Issues with doing `&&` 
* `postinstall` script is too long
* It is runs sequentially which can get very very slow

We can install `frontend/` and `backend/` dependencies in parallel to make it quick. That is why we use [`npm-run-all`](https://www.npmjs.com/package/npm-run-all) and this also makes the install script look as one script.

Suppose you clone the project and use `npm install-deps` that command will crash because it depends on `npm-run-all` and this is why I suggested to remove that from docs because as a developer you don't have to run it ever manually because the way this project is structured.

Reply:
What I'm getting at is "If the command is broken, why is it in package.json"?

Reply:
Am I missing something in how this command functions? If it's not for the user, which other script can use it if `npm-run-all` isn't installed?

Reply:
And if the command should not exist in package.json, shouldn't someone open a new pr and fix that?

Reply:
> Am I missing something in how this command functions?

Yes

> If it's not for the user, which other script can use it if npm-run-all isn't installed?

`npm-run-all` will be installed when you do `npm i` at `.` because that is what the main thing one needs to do after cloning this project.

> And if the command should not exist in package.json, shouldn't someone open a new pr and fix that?

No the command will work. The only situation it will be broken if you DO NOT `npm install` at `.`

This is a helper command that will automatically run once you do `npm install` in `.` and there is no situation I can think in which a developer will run `npm run install-deps` so it is fine not being documented. 

You can check mobile, presenter repo's  `package.json` scripts section we have many "helper" scripts that we use to help organize and not all are required to be documented. If someone is interested they can do `npm run` in terminal or checkout scripts section in `package.json` for implementation.

Reply:
> # Run npm install for entire project

This can be misleading since that command will only install dependencies for `frontend/` and `backend/` and we make an assumption that they already have top level dependencies installed or have `npm-run-all` globally installed